src/s3.rs

```diff
@@ -54,6 +59,51 @@ pub fn get_client(shared_config: &aws_config::Config) -> Result<Client> {
     Ok(Client::from_conf(builder.build()))
 }
 
+#[derive(Debug)]
+pub enum ListObjectsError {
+    ListError(SdkError<ListObjectsV2Error>),
+    MissingContents(ListObjectsV2Output),
+}
+impl fmt::Display for ListObjectsError {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        match self {
+            Self::ListError(e) => write!(f, "Error while listing objects: {}", e),
+            Self::MissingContents(e) => {
+                write!(f, "List objects output missing contents: {:?}", e)
+            }
+        }
+    }
+}
+
+/// Perform a bucket listing, returning a stream of results.
+///
+/// # Example
+///
+/// ```
+/// FIXME
+/// ```
+///
+pub fn list_objects(
+    client: &Client,
+    bucket: &str,
+    prefix: Option<String>,
+) -> impl Stream<Item = Result<model::Object, ListObjectsError>> + Unpin {
+    let req = client
+        .list_objects_v2()
+        .bucket(bucket)
+        .set_prefix(prefix)
+        .into_paginator();
+    req.send()
+        .map_err(|e| ListObjectsError::ListError(e))
+        .map_ok(|list_objs| {
+            stream::iter(match list_objs.contents {
+                Some(c) => c.into_iter().map(Ok).collect(),
+                None => vec![Err(ListObjectsError::MissingContents(list_objs))],
```

Comment:
It's a shame to have to `collect` here. Is this so that the two match arms can easily return something of the same type?

Reply:
Yeah, that's exactly it. There's perhaps an alternate pattern for this whole block that achieves the same thing? Would be happy to take alternate spellings of this to avoid the collect.

Reply:
I'm currently staring at a similar situation and using `Box`, but I'm not sure it's much better:

```rust
    pub fn dicoms_to_reassemble<'a>(
        &'a self,
        client: &'a s3::Client,
    ) -> impl Stream<Item = Result<String>> + 'a {
        self.list(client, "reassembly-manifest/dicoms")
            .and_then(move |dicoms_file| async move {
                let items: Box<dyn Stream<Item = Result<String>> + Unpin>;
                items = match self.read(client, &dicoms_file).await {
                    Err(e) => Box::new(stream::once(future::ready(Err(e)))),
                    Ok(body) => Box::new(
                        BufReader::new(GzipDecoder::new(body.into_async_read())).lines().map_err(anyhow::Error::from)
                    ),
                };
                Ok(items)
            })
            .try_flatten()
    }
```

(The key part is the `let items`, which lets you store the stream on the heap and hide its concrete type).